---
title: "Tutorial Basics (and some trics) for tree annotation with ggtree"
author: "Eve Zeyl Fiskebeck"
date: "1/3/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Principles of trees annotations with `ggtree`

This tutorial is a quick start to help you annotating trees with `ggtree` package 
using in R. `ggtree` library is based on `ggplot2` library, so if what you want to
do is not described specifically in ggtree, have a look of how to do it with ggplot2,
this is very useful.

Please also refer to the reference books:
- [Data integration, Manipulation and Visualization of phylogenetic trees](https://yulab-smu.top/treedata-book/index.html), 
hereafter refereed as "ggtree book". 
- [ggplot2: Elegant Graphics for Data Analysis](https://ggplot2-book.org/), hereafter refereed as "ggplot2 book". 
- [Analysis of Phylogeneitcs and evolution with R](https://www.mimuw.edu.pl/~lukaskoz/teaching/sad2/books/Analysis_of_Phylogenetics_and_Evolution_with_R.pdf), a reference
book that provide description of how to do tree manipulations with `ape` R library. 
- [R for data science](https://r4ds.had.co.nz/), quick start for data import, manipulation and visualization. 

Note: `ggtree` library, and other libraries eg. `treeio` described in "ggtree book" are
being actively developed. It might occur that the documentation is not allways in 
sync with the library version that you have installed (ie. functions you do not find anymore,
functions where syntax has been modified.) In this case: look at the functions 
available in the library versions you have loaded, and if this does not help there 
is usually solutions to be found in [stack overflow](https://stackoverflow.com/). 

## Tutorial setup: libraries installation

`ggtree` library can be installed through the [Bioconductor](https://bioconductor.org/) package manager, 
this is explained [here](https://bioconductor.org/packages/release/bioc/html/ggtree.html. 

```{r install ggtree, eval=F, echo=TRUE}
# source: https://bioconductor.org/packages/release/bioc/html/ggtree.html
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ggtree")
```

Installation of other libraries we will use in this tutorial:

The `pacman` library is convenient to install very fast new libraries (less writting)
and is also usefull to explore the functions in each library.  
```{r install pacman, eval=F, echo=TRUE}
install.packages("pacman", dependencies = T)
```

Loading and/or installing other libraries with pacman:
```{r loading required packages, eval=F, echo=TRUE}
p_load(dplyr, ggtree, treeio, ape, ggplot2, ggnewscale)
```

- [ ] need check if treeio in bioconductor or ?

Finding out what functions are available within a specific package
```{r using pacman to find functions in a specific package, eval=F, echo=TRUE}
p_functions(ggtree)
```

## In short: Principles of graphics with `ggplot` and libraries based on ggplot
`ggtree` makes use the grammar of graphics, as described in "ggplot2 book". 

In short the plot is composed of:
- the data (the tree and associated metadata)
- the mapping parameters: how the data variables (eg. geographical origin, age, ID, bootstrap values) are mapped to an aesthetic attribute
- several layers. Layers can be seen as a set of transparent sheets, that you add on top of each other, each adding a different information, and that together will create the visual effect that you wished to accomplish (hopefully). 
- scales: for eg. labeling colors, points size, shape and fill. Scales are either discrete or continuous.
- a coordinate system (for trees: it depends on the length of your tree)
- facets (used to break up the data set in subsets for display of mutliple subsets - we wont use that here)
- a theme (which defines eg. the background color or your plot, font familly, location/display parameters of the legend... we will use very, very simple theme)

We will sequentially define all those elements to be able to visualize and annotate a phylogenetic tree. 
For more information, please have a look at the "ggplot2 book"

## The data used in this tutorial

We will use data from the following article [Cohen et al. 2021](https://doi.org/10.3389/fmicb.2021.729637) 
that uses comparative analysis and phylogenetic lineages reconstruction to study the differences in distribution of _Actinobacillus pleuropneumoniae_ AMR genes between Norway and Denmark. 

We have downloaded the sequences of the isolates used in this article and associated metadata. 
We have reconstructed a ML phylogenetic tree (for isolates belonging to clade 2) using  [ALPPACA](https://github.com/NorwegianVeterinaryInstitute/ALPPACA) pipeline.

The ML tree inference is performed with [IQTREE](http://www.iqtree.org/). 
We will use the consensus tree (1000 boostrap) for this tutorial.
The tree is provided in [newick format](https://en.wikipedia.org/wiki/Newick_format#:~:text=In%20mathematics%2C%20Newick%20tree%20format,Maddison%2C%20Christopher%20Meacham%2C%20F.)

Data location on SAGA `path`
- metadata file: `metadata_clade2_tutorial.rds`
- consensus tree file: `IQTREE_tree.phylo`(the default output name for your tree when using ALPPACA)
- pairwise SNPs distance: `distance_cluster2_clean.rds` long format, no double pairs, no isolate distance to itself*


# Tutorial

Setup: see above

## Importing data in R

The metadata have been saved in a _single R object_ file to simplify import for this tutorial. 

To be able to use the metadata for tree annotation, the variables have to be of the correct type
(what should be a real should be defined as such, and characters / factors as such).

Improper data types for your variables typically produce scales error (ie discrete value provided to a continuous scales or vice-versa),
and also eg. can hinder recognizing adhecally the variable you want to display... 
So you need to have an idea of how you want your tree to look to be able to know which type is appropriate 
Here is an example how to change the types. 

1. We read the metadata in an object (change the <path>)
```{r assigning r object into memory, eval=F, echo=TRUE}
metadata_file <- "<path>/ggtree_tree_annotation/metadata_clade2_tutorial.rds"
metadata <- readRDS(metadata_file)
```

2. Check if all variables types are in correct format:
```{r summary of object structure, eval=F, echo=TRUE}
str(metadata)
```
3. We change the data type to data type that we want
I want the year to be display as a continuous gradient so I can se a natural evolution, I need to change the type from integer to real (double in R). The ID for isolates I will use, and are ok, they are as character. I want the Farm_ID to be a character (for labeling).
> ggplot often transforms alone between datatypes (but its good practice to ensure that they are adequate)

```{r mofifying the types of metadata variables, eval=F, echo=TRUE}
metadata <- metadata %>%
  mutate_at(.vars = "year", as.double) %>%
  mutate_at(.vars = "Farm_ID", as.character)

str(metadata)
```

Year is now of numeric (real) type instead of integer. I will be able to use a continuous scale. 

4. We import the tree into memory - check the tree
```{r mofifying the types of metadata variables, eval=F, echo=TRUE}
tree_file <- "<path>/IQTREE_tree.phylo"
tree <- ape::read.tree(tree_file)
tree
str(tree)
```

The tree is unrooted (it is a ML tree!). 
We see that the sequences names are used as Tip lables. Because of way [Parsnp](https://harvest.readthedocs.io/en/latest/content/parsnp.html) is used in ALPPACA (we asked to choose a random sequence as reference), we have one dplicate sample annotated as `xxxx.fasta.ref` and `xxxx.fasta`. This is not a problem for the tree reconstruction with IQTREE as IQTREE remove duplicates sequence from the tree inference and reinsert them afterwards. But it is not nice to have an isolate displayed twice on a tree. 

## Preliminary: preparing data for display 

There for we detect which isolate is the reference that is counted in double, and drop this tip. 
```{r drop reference tip: isolate in double, eval=F, echo=TRUE}
# detect tip to drop
drop_tip <- tree$tip.label[stringr::str_detect(tree$tip.label, ".fasta.ref")]
# new tree with droped tip
tree <- treeio::drop.tip(tree, drop_tip)
```

We clean the tip label names to be able to match and combine with the metadata IDs to the tree 
```{r drop reference: isolate in double, eval=F, echo=TRUE}
tree$tip.label <- str_remove(tree$tip.label, ".fasta") 
tree$tip.label
```

We select the metadata that only correspond to the tree (optional, but this helps avoiding NA values when plotting).

```{r metadata specific to the tree, eval=F, echo=TRUE}
metadata_tree <- 
  tibble(id = tree$tip.label) %>% # creates a dataframe with the tips ids from the tree
  left_join(metadata, by = c("id" = "Acession_number")) %>% # joining 2 data frames, specifying which columns are corresponding keys
  select(id, everything()) # reorder such as id is the first column (required to associate tree and metadata in plotting)
```

## From plotting a simple tree to a more advanced tree
1. Specify data used in the tree and make a simple plot
```{r specify data tree, eval=F, echo=TRUE}
simple_plot <- ggtree::ggtree(tree, 
                 layout = "rectangular", 
                 ladderize = T, 
                 lwd = .2) %<+%
  metadata_tree 
simple_plot
```

2. add layers to display points and labels and bootstraps
We have created a plot object. This plot object contains the instructions on how to make a plot (a blueprint). The plot is not done
(instructions not evaluated) until I ask the render. The evaluation happens when I type simple_plot. This is very useful because 
I allows me to experience and add gradually more instructions to my blueprint, which allow to test if it does what I want before assigning that to an object. 

We add tip labels, and color by year (in the aesthetic!) and point shape for region.

Note: Bootstrap values are normally associated to the branch preceding a node, and represents the support for a specific clade (including all descendants). However, several software used to display bootstrap values, display those values at the nodes. `ggtree` does not have a function to display bootstrap on branches but we can display those at nodes, and adjust the position of the labels accordingly. 

The scale bar provide the correspondence between the evolutionary distance (the mean number of substitutions per site) for a branch of corresponding length. 

```{r specify mapping and layers: geoms color, shape, labels tips and nodes, eval=F, echo=TRUE}
simple_plot 

simple_plot + 
  geom_tiplab(aes(label = label, color = year), size = 2) +
  geom_tippoint(aes(shape = Region)) +
  geom_nodelab(size = 2, nudge_x = -0.0000009, nudge_y = 0.5) +
  geom_treescale(fontsize = 2) 
```

### Aparte: 
NB: There is no good solution to find how much you should adjust the positions. It depends on the size of your tree. 
However, you can find what is the range of the xlim and ylim as soon as you build a simple plot, this can give you an idea of
the range of value and then on what order of magnitude you need to use for adjustment:

```{r finding size of the tree, eval=F, echo=TRUE}
# Most useful
myplot_xlim <-ggplot_build(simple_plot)$layout$panel_scales_x[[1]]$range$range
myplot_xlim
# Just in case ...
myplot_ylim <- ggplot_build(simple_plot)$layout$panel_scales_y[[1]]$range$range
```

```{r visualizing xlim and ylim, eval=F, echo=TRUE}
simple_plot + 
  geom_vline(xintercept = myplot_xlim[1], color = "green", lty = 2) +
  geom_vline(xintercept = myplot_xlim[2], color = "green", lty = 3) +
  geom_hline(yintercept = myplot_ylim[1], color = "blue", lty = 2) +
  geom_hline(yintercept = myplot_ylim[2], color = "blue", lty = 3) 
```


### Improving the tree visualization

I am not happy with the display of the regions. I want only full symbols filled with color. 
Shapes are categorical variable, and we want to specify EXACTELY how they look like. 
To do so we define a palette, that defines the correspondence between a shape and a value of the categorical variable. 

Shapes are defined by the graphical parameter pch 

```{r example of shape palette, eval=F, echo=TRUE}
# find what the shapes correspond to
?pch 
# Create a shape palette 
palette_shape <- c("Central"= 21,
                   "East" = 22,
                   "Greater Oslo" = 23,
                   "North" = 24,
                   "South-West" = 25)
```

```{r improve shape, eval=F, echo=TRUE}
simple_plot + 
  geom_tippoint(aes(shape = Region)) + 
  scale_shape_manual(values = palette_shape, name = "Shape = Region") 
```
This looks good, so now I want also the color and fill corresponding to the Farm 

```{r improve shape, eval=F, echo=TRUE}
simple_plot2 <- simple_plot + 
  geom_tippoint(aes(shape = Region, fill = Farm_ID, color = Farm_ID), size = 2, show.legend = T) + 
  scale_shape_manual(values = palette_shape, name = "Shape = Region") 
simple_plot2
```

I tip labels, and I want them to be colored according to the year of sampling. 
In a normal ggplot you can only have one scale per type display (shape, color,
size, fill). The library `ggnewscale`allow us to create new scales, but this means 
that all the drawing parameters from the previous scale, need to be defined before defining
a new scale. 

```{r add a new scale and colored tips labels, eval=F, echo=TRUE}
simple_plot2 +
    ggnewscale::new_scale_color() +
    geom_tiplab(aes(label = label, color = year), size = 2) 
```

Now I want the labels to be the samples IDs not accession numbers and also I want 
the farms ID as labels to ease interpretation. And I do not like this gradient of color for the year, so 
I will specify it. Here I use a continuous color scale (gradient, and specify the middle value for the color value)

```{r improve scale and color scale for colored tips labels, eval=F, echo=TRUE}
simple_plot3 <- simple_plot2 +
  ggnewscale::new_scale_color() +
  geom_tiplab(aes(label = paste0(ID, " F ", Farm_ID), color = year), size = 2) +
  scale_color_gradient2(midpoint = 2011.5, low = "orange", mid = "red",
                        high = "blue", space = "Lab", name ="Year = sampleID colored labels") 

simple_plot3
```
Now we add the bootstrap values, and colored in red if they are low (<95 for fast bootstrap with IQTREE) and green otherwise.

```{r add the conditional bootstrap labels, eval=F, echo=TRUE}
simple_plot4 <- 
  simple_plot3 + 
  ggnewscale::new_scale_color() +
  geom_nodelab(aes(color = as.numeric(label) < 95),
               node = "internal", 
               size = 1.5, nudge_x = -0.0000009, nudge_y = 0.5) +
  scale_color_manual(values = setNames(c("red", "darkgreen"), c(T,F)),
                     na.translate = FALSE, name = "bootstrap") +
  geom_treescale(fontsize = 2)
simple_plot4
```
Now we would like to add title, move the legend to the other size.

```{r change the position of legend add title, eval=F, echo=TRUE}
simple_plot5 <- 
  simple_plot4 +
  guides(color = "none") + # This allows removing the boostrap legend
  ggtitle("Clade 2") +
  theme(legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        plot.title = element_text(),
        legend.position = "left")
simple_plot5
```

### Saving you graph

Be aware that the appearance of your graph does not only depend on the parameters
used for display but ALSO of the graphical display and its format!. 

So if your graph is a bit squished in width or in height, sometimes this can be
adjusted by changing the settings of width and height for import. But this also 
can go into opposites direction, if a plot looks nice on the console display, 
it can become very ugly if you change the ratio of the display when saving.

```{r change the position of legend add title, eval=F, echo=TRUE}
ggsave("my_colorfull_tree.png",
       plot = simple_plot5,
       units = "cm", dpi = 600, width = 30, height = 25, type = "cairo")
```

### Annotating clades on the with distances.

There is unfortunately no easy (automatic) way to do so as per today.
To do so you need to:

1. Get the taxa names (best to get them in order of the tree). 

Note: Be aware the taxa names are the tip names in order as represented in the tree, 
but the labels are the tree labels, not the metadata ID labels we used as display.
Redo a simple tree plot using `geom_tiplab(aes(label = label), size = 2)`
to be able to view the corespondance. 

```{r get taxa names, eval=F, echo=TRUE}
taxa_names <- get_taxa_name(simple_plot5)
```

2. Now locate the high and low taxon that are contained in the clade of interest, and
extract the corresponding node number
```{r get taxa names, eval=F, echo=TRUE}
group1_node <- MRCA(tree, "ERR6510225", "ERR6510223")
group1_node 
```

3. subset your taxa name vector to extract all the taxa that belong to the group of interest

```{r get taxa names, eval=F, echo=TRUE}
# you can do manually
group1 <- taxa_names[12:16]
# or:
group1 <- taxa_names[which(taxa_names %in% "ERR6510225"):which(taxa_names %in% "ERR6510223")]
```

4. Extract the pairwise distance for the group in question

```{r extract pairwise distance for selected group, eval=F, echo=TRUE}
# load all the pairwise distances
distance_cluster2 <- readRDS("distance_cluster2_clean.rds") 
group1_distance <- distance_cluster2 %>%
  filter(ID1 %in%group1, ID2 %in% group1) %>%
  pull(distance) %>%
  range() # here we give the range (min and max)
```

5. Further annotate the tree. 1 Annotation per group
```{r extract pairwise distance for selected group, eval=F, echo=TRUE}
simple_plot5 +
  geom_cladelab(node = group1_node, label = "0-1\nSNP", 
                align = F, geom = 'text', 
                textcolor = "black", size = 1, offset =.00001)
```

Use one line of `geom_cladelab` per group you want to annotate.

If your plotting area is too small, you can extend the xlim (or ylim, depending 
on you type of annotation) as done bellow.

If you extend too much, your tree will be all squished, so you need to get back to 
the limits of the tree in the plot and then adjust accordingly. 


```{r trick: extend you plot area, eval=F, echo=TRUE}
# Extract the current xlim (as done above)
ggplot_build(simple_plot5)$layout$panel_scales_x[[1]]$range$range

# redefine (slighly extend) xlim
simple_plot5 + 
    geom_cladelab(node = group1_node, label = "0-1\nSNP", 
                align = F, geom = 'text', 
                textcolor = "black", size = 1, offset =.00001) +
  xlim(0.000000e+00, 5e-05)
```



# In need of something specific ? 
WARNING: be aware that if you import you tree as treedata (treeio::read_iqtree) and need to reroot your tree this will lead to incorrect 
bootstrap values. Therefore it is better to directly use ape to import the tree, and rerooting should be the first step you do (before merging your data) 
- [ ] put somewhere how to reroot the tree. 


Please ask us if you are in need of certain types of displays.
We have written scripts for different projects, that could be adapted and reused.



You can find our examples of code here:

- [APEC-2021-Rprepackage](https://github.com/NorwegianVeterinaryInstitute/APEC-2021-Rprepackage)
> contain scripts for cleaning and working with distance data (transforming to clean long format)*. Of special interest.
  - transforming SNPs distances and extracting some summary statistics and distance plots
  - selecting metadata and merging with tree data (in order to avoid NA and further problems when plotting)
  - functions for proper rerooting of trees using ggtree  
  - tree annotation - tree/rotated_tree/distance heatmap plot in one graph (allow to visualize pairwise distance between groups)

- [VI31142_NRC250212_Broiler](https://github.com/NorwegianVeterinaryInstitute/VI31142_NRC250212_Broiler)
> contains scripts for plotting snippy and beast phylogenetic trees, different scales
 - contain scripts to be able to subset trees (allows "zooming" on specific clade -> to make a zoomed plot)
 
- [yersinia](https://github.com/NorwegianVeterinaryInstitute/yersinia/tree/master/src)
> contains scripts for phylogeography
  - plotting different trees on a map, and within a timeline (facetplots)
  
- [ggevxtra](https://github.com/NorwegianVeterinaryInstitute/ggtrevextra)
> planned package (in prep) that will take the tools above in coherent package

# Going further
Shapes: 
  - [More Shapes](http://sape.inf.usi.ch/quick-reference/ggplot2/shape)

  
# References

Cohen, Liza Miriam, Janine T. Bossé, Marc Stegger, Yanwen Li, Paul R. Langford, Camilla Kielland, Thea Blystad Klem, et al. “Comparative Genome Sequence Analysis of Actinobacillus Pleuropneumoniae Serovar 8 Isolates From Norway, Denmark, and the United Kingdom Indicates Distinct Phylogenetic Lineages and Differences in Distribution of Antimicrobial Resistance Genes.” Frontiers in Microbiology 12 (2021): 2558. 
https://www.frontiersin.org/articles/10.3389/fmicb.2021.729637/full



Packages:
- ggtree
- ape
- pacman
- ggnewscale 
- ggplot2










